### PR TITLE
thread safe metric initialization

### PR DESCRIPTION
**Pull request by gcolvin**
Attempted fix of potential race condition initializing metrics structure.  Since no known failure has been observed it's probably not testable, so please review the two lines carefully.

### CODE REVIEW

File: libevm/VMOpt.cpp [hunk 1]
```diff
@@ -35,9 +35,9 @@ void VM::reportStackUse()
 std::array<InstructionMetric, 256> VM::c_metrics;
 void VM::initMetrics()
 {
-	// FIXME: This is not thread safe.
-	static bool done=false;
-	if (!done)
+	// initialization of statics is thread safe, so exchange() will return false only once
+	static std::atomic<bool> done(false);
+	if (!done.exchange(true))
```
Comment (chfast) [inline libevm/VMOpt.cpp]: I think this is worse now. If the first still writes to c_metrics the second will see that the flag is true, will skip the initialization part and use c_metrics partly initialized.

To do it in "easy" C++11 way, you can put all initialization into a lambada returning true, and assign this lambda invocation to a `static bool initialized` variable. C++11 guarantees thread-safe static initialization.

See this example: https://github.com/ethereum/cpp-ethereum/blob/develop/libdevcrypto/LibSnark.cpp#L40-L47

Looks hackish, but this allows the compiler to choose the best synchronization option available.

Comment (gcolvin) [reply libevm/VMOpt.cpp]: You are right, thanks. I wrote it with the lambda first (which is really cool and requires no library help) but then worried whether everything going into the initialization was itself statically initialized.  So I did something worse.  Chasing back the instructionInfo() function I see no problem, so I'll go back to my first approach.

Comment (gcolvin) [reply libevm/VMOpt.cpp]: Digging a little deeper I'm trying to clean up the unnecessarily cluttered initialization code.

**Comment by gcolvin** [comment]
@chfast I think this ready.

**Pull request merged**